web-automation-agent/src/shrink_html.ts

```diff
@@ -0,0 +1,54 @@
+import cheerio from 'cheerio';
+import { type Page } from 'puppeteer';
+import { WHITELIST_ATTRIBUTES_WEB_AUTOMATION, WHITELIST_TAGS_WEB_AUTOMATION } from './consts.js';
+
+interface ShrinkHtmlOptions {
+    whiteListTags: string[];
+    whiteListAttributes: string[];
+    maxElementTextLength?: number; // TODO: Implement this
+}
+
+/**
+ * Tag each element in the HTML with a unique attribute.
+ */
+export async function tagAllElementsOnPage(page: Page, attributeName: string) {
+    return page.$$eval('html *', (elements, attrName) => {
+        for (let i = 1; i < elements.length; i++) {
+            if (!elements[i].getAttribute(attrName)) elements[i].setAttribute(attrName, `${i}`);
+        }
+    }, attributeName);
+}
+
+export async function shrinkHtml(page: Page, options: ShrinkHtmlOptions) {
```

Comment:
We could also try to remove all the `<div>`s or at least the `<div>`s that are directly in the other `<div>`. As it does not add any semantical context:

```
<div>
  <div>
    <div>
      something
    </div>
  </div>
</div>
```

Reply:
The same way for

```
<div>
  <p>...</p>
</div>
```

as the only paragraph is in the div, the div itself does not add any structural context

Reply:
Maybe it could even work better without all the divs, who knows